CONTRIBUTING.md

```diff
@@ -24,4 +24,70 @@ The core developers will review the pull request and may suggest changes.  Simpl
 changes to the branch that is being pulled from, and they will automatically be added to the
 pull request.  In addition, the full test suite is automatically run on every pull request,
 and rerun every time a change is added.  Once the tests are passing and everyone is satisfied
-with the code, the pull request will be merged.  Congratulations on a succesful contribution!
\ No newline at end of file
+with the code, the pull request will be merged.  Congratulations on a successful contribution!
+
+## Building OpenMM
+
+OpenMM uses [cmake](https://cmake.org/) as our build system and requires quite a number of dependencies.
+
+### Installing dependencies
+
+OpenMM's dependencies are described in detail in [the User Guide](http://docs.openmm.org/latest/userguide/library.html#compiling-openmm-from-source-code). The easiest and fastest way to install all required dependencies is to use the [Conda package manager](https://conda.io).
+
+#### on Linux
+
+```shell
+# Install build requirements
+conda env create -n openmm-dev -f devtools/ci/gh-actions/conda-envs/build-ubuntu-latest.yml
```

Comment:
> I think the following command will install everything they need from conda (but please check it to be sure!)
> conda install -c conda-forge cmake cython swig fftw doxygen=1.8.14

The whole point of making it easy for users to install development environments is to avoid having them manually install a bunch of dependencies. At the very least, we should provide a current conda YAML environment that can be used to create a working build environment. Even better would be to also push this environment to conda-forge and keep it current so you can simply specify the build environment to be installed from conda-forge, like [qcfractal does](http://docs.qcarchive.molssi.org/projects/QCFractal/en/stable/install.html#conda-pre-created-environments).

Reply:
Providing an environment like this would also harmonize dependency installation across all supported platforms.

Reply:
We want to make it as easy as possible for people to set up a useful development environment. If we somehow update the manual to have the simplest, clearest achievable way of doing this, we can just point to it. But [the current installation instructions](http://docs.openmm.org/latest/userguide/library.html#compiling-openmm-from-source-code) is a loooooong way from this.

Reply:
We could create another yaml file to install dependencies needed for development.  That still wouldn't give a full development environment, though.  The unfortunate truth is that compiling OpenMM is complicated.  You still need an SDK for OpenCL or CUDA.  It will be a different one depending on whether you have an NVIDIA, AMD, or Intel GPU, and none of them is conda installable.  (The "cudatoolkit" package from conda-forge isn't really the full toolkit, just the redistributable libraries.)  And you need an appropriate compiler.  And unfortunately, the approach used by conda-forge of installing their own versions of system libraries right down to libc++ complicates things further.  If you want to compile OpenMM for use in Python along with packages from conda-forge, you need to compile against their libraries.  But if you want to compile it as a library for use in C++/C/Fortran programs, you need to compile against the OS versions of the libraries.

Reply:
@peastman : Do we know what fraction of folks develop for the Python API or develop for linking with C++/C/Fortran? If not, should I to a brief survey on twitter?

My expectation is that most of our users simply need a conda-installable dev environment that is close to the build environment for conda-forge to do some tinkering with the Python layer (or a bit of the C++ layer). Power users may want to set up their own development environment to use with linking with C++/C/Fortran. If we can separate these two classes, we will have something simple for most users and something detailed for power users.

Reply:
No idea what the fractions are.  If someone is compiling from source, I think that automatically puts them into the power user category.  The question of course is *why* they're compiling from source.  Either because they plan to modify it, or because they need to create libraries for an unusual environment (one not supported by conda-forge).

Reply:
Or of course a third possibility: they need a version that works with standard OS libraries, not the ones from conda-forge.

Reply:
Note that CONTRIBUTING.md are instructions for potential contributors, not users. Users should install from conda-forge (directions should be in the readme). But we also need simple, clear instructions that will take someone familiar with git and the CLI from zero to contributing, or allow an experienced contributor to set up their dev environment quickly on a fresh machine. Potential contributors *will* be put off by having to read an entire chapter of the manual to find three lines at the terminal, especially if they are incorrect (which they are without this PR and most likely will be again unless they're tested in CI). I know this because *I* am finding it off-putting, and I'm being paid!

Conda environments are *great* for this, *especially* Conda environments that are tested in CI. With one command you get all the dependencies, including compilers and CUDA/OpenCL and the testing infrastructure, and when you're done working on OpenMM you just delete the environment and don't have to worry about messing up dependencies for other projects.

This is especially important for OpenMM because:

1. OpenMM is complex (as are all MD codes), and correct tests and up-to-date documentation are especially important
2. Documentation is a great way for new contributors to get their feet wet, especially those that aren't expert programmers
3. Lots of expert Python coders do not have C++ compilers installed
4. We need a *C++ compiler* to build not only the *Python wrapper* but also the *Python API docs*. I'm not sure many dependencies would be removed if we were only targeting Python devs.

I think it would be amazing to add a new Conda `environment.yaml` file that has all the build dependencies, including compiler, testing stuff, documentation, and OpenCL/CUDA, as long as we test that it builds everything correctly in CI.

Reply:
Totally agree with everything @Yoshanuikabundi notes here!

Reply:
e3fb262 and 60d7d21 adds `devtools/environment.yml`, which includes all the dependencies needed to compile OpenMM and the HTML docs, but does not include LaTeX or any GPU SDKs. I've also updated `CONTRIBUTING.md` to focus on getting people compiling the CPU platform quickly, but also give some clues on choosing a GPU SDK from conda-forge. I figure this probably captures a large amount of what contributors would be working on without requiring us to get into the nitty gritty of detailed instructions for every combination of hardware out there. Finally, I've added CI to make sure the instructions continue to work. The given instructions are solid for the CPU platform on MacOS and Linux. I've had to comment out the Windows matrix entry for the `CONTRIBUTING.md` docs (not for any existing CI) because the ordinary unix commands like `make` don't seem to work.

Reply:
> With one command you get all the dependencies, including compilers and CUDA/OpenCL and the testing infrastructure

I wish that were true!  Unfortunately, you don't really get a fully functional environment.  You also need to install the CUDA and/or OpenCL SDK appropriate for your GPU.  Those are not conda installable.  (The "cudatoolkit" package is not really the full toolkit, only the redistributable libraries.)  Also, this procedure will often produce libraries that can *only* be used within the conda environment.  That's because conda-forge ships its own versions for lots of system libraries.  If someone is building libraries to use from a C++ or Fortran program, there's a good chance they'll incompatible libraries that don't work with the standard system libraries built into the OS.

I think we really need to document two separate procedures, depending on whether you're building packages to use inside or outside a conda environment.

Reply:
Please see the latest commit in this PR, where the proposed environment file is implemented and CONTRIBUTING.md has been updated to provide instructions on its use.

>  You also need to install the CUDA and/or OpenCL SDK appropriate for your GPU. Those are not conda installable. 

Actually they are; see the feedstock where the conda distributed binaries are built from conda dependencies only. The trick is that there isn't a one size fits all solution, different hardware needs different packages. In the latest commit to this PR, I explicitly discuss this, and note that the environment is only sufficient to build the CPU platform. I think this will cover a lot of development needs, and I also provide hints for finding the appropriate GPU sdk package for contributors who need it.

> The "cudatoolkit" package is not really the full toolkit, only the redistributable libraries.

`cudatoolkit-dev` includes the headers.

>  Also, this procedure will often produce libraries that can only be used within the conda environment. That's because conda-forge ships its own versions for lots of system libraries. If someone is building libraries to use from a C++ or Fortran program, there's a good chance they'll incompatible libraries that don't work with the standard system libraries built into the OS.

True, but this is a feature, not a bug, as it allows devs to maintain the entire dependency chain in a single conda environment rather than having libraries strewn across their system. I'd argue that this is actually the entire point of using conda! 

> I think we really need to document two separate procedures, depending on whether you're building packages to use inside or outside a conda environment.

This PR doesn't remove any documentation. The existing build guide in the manual still exists. I can add a line to contributing.md mentioning this caveat if you like?

Reply:
> Please see the latest commit in this PR

That's looking a lot better!  Just a few more comments.

> cudatoolkit-dev includes the headers.

Even that still isn't sufficient to build it, unfortunately.  To build the conda packages, we need to use special Docker images that have the necessary (non-redistributable) components pre-installed in them.  For example,

https://github.com/conda-forge/openmm-feedstock/blob/master/.azure-pipelines/azure-pipelines-linux.yml#L14

I don't know of any way to get a complete, functional CUDA build environment that doesn't involve obtaining packages directly from NVIDIA.

> True, but this is a feature, not a bug

That depends on what you're doing! :)  If you're a Python developer building a library to use inside a conda environment, it's a feature.  If you're a C++ developer building a library to link with a stand-alone application you're writing, it's a bug.

Reply:
It seems like we're dealing with two use cases here, then. Why not split this into two paths, especially since the first (Python) path is likely to be quite popular and very easy/streamlined?

Reply:
Also, maybe you have ideas about how to streamline the installation of a dev environment for the second case, @peastman?

Reply:
The paths wouldn't be Python/Not Python, they'd be GPU platforms/everything else. But yes! Different use cases.

> I don't know of any way to get a complete, functional CUDA build environment that doesn't involve obtaining packages directly from NVIDIA.

`cudatoolkit-dev` just pulls the SDK installer down from NVIDIA and executes it. I'm installing it now, so I guess we'll find out if it works shortly!